source/vstd/std_specs/hash.rs

```diff
@@ -0,0 +1,383 @@
+/// This code adds specifications for the standard-library type
+/// `std::collections::HashMap`.
+///
+/// Most of the specification only applies if you use `HashMap<Key,
+/// Value>`. If you use some custom build hasher, e.g.,
+/// with`HashMap<Key, Value, CustomBuildHasher>`, the specification
+/// won't specify much.
+///
+/// Likewise, the specification is only meaningful when you know that
+/// the `Key` has a deterministic hash. We have an axiom that all
+/// primitive types and `Box`es thereof have a deterministic hash. But
+/// if you want to use some other key type `MyKey`, you need to
+/// explicitly state your assumption that it has a deterministic hash
+/// with
+/// `assume(vstd::std_specs::hash::does_type_have_deterministic_hash::<MyKey>());`.
```

Comment:
Can we shorten `does_type_have_deterministic_hash` to `deterministic_hash` and `does_type_conform_to_build_hasher_model` to something like `obeys_hasher_model`?

Reply:
Given what you said above, we need something stronger than that it has a deterministic hash. It has to do that and also require keys satisfying `==` to be identical. How about `satisfies_hash_table_key_model`? As for `obeys_hasher_model`, it's not that the type has to obey our model of `Hasher`, it's that it has to construct objects that obey that model. How about `satisfies_build_hasher_model`?

Reply:
> also require keys satisfying == to be identical

Would it be enough to require identical views, rather than identical keys?

>  How about `obeys_hash_table_key_model`?

Since we're already in the HashMap namespace, perhaps `obeys_key_model` would suffice?

> How about `obeys_build_hasher_model`?

That could work.  What do you think of `builds_valid_hashers`?

Reply:
I've updated the names to `obeys_key_model` and `builds_valid_hashers`.

Reply:
Cool, sounds good.